### PR TITLE
!!! FEATURE: Add Afx template for SiteKickstarter

### PR DESCRIPTION
**Pull request by Wyndala**
!!! FEATURE: Add Afx template for SiteKickstarter

This is the first approach for making the SiteKickstarter more versatile. Now you can easily add a custom generator package and kickstart some sorts of sites with one package.

* Add possibility to use more than one template for the site kickstarter
* Add Afx template in addition to the fluid template
* Add the generator selection to the backend module for site packages
* Removed the old generic generator service (this is breaking)

This is a breaking change because it removes the old, Fluid-based generator service.
In case you want to use the Fluid based rendering you can use the dedicated package at https://github.com/queoGmbH/FluidSiteGenerator

Resolves: #3074

### CODE REVIEW

**Comment by bwaidelich** [comment]
Awesome, thanks for this. I need to look into this more closely, but I already added it to the [6.0 project](https://github.com/orgs/neos/projects/24) so that we can hopefully integrate it with the next major

**Comment by bwaidelich** [comment]
Update: I just realized this was missing a corresponding feature ticket. I created one (#3099), linked it and added _that_ to the project board :)

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php
Comment (daniellienert) [inline Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:256]: As this changes the interface already, please add argument and return types

File: Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php:1]:
```suggestion
<?php
declare(strict_types=1);
```

File: Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php:6]: Please remove that comments

File: Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php:4]: Why is this comment removed?

File: Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php:1]:
```suggestion
<?php
declare(strict_types=1);
```

File: Neos.SiteKickstarter/Classes/Generator/AbstractSitePackageGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/AbstractSitePackageGenerator.php:7]: Please add argument and return types

File: Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:1]:
```suggestion
<?php
declare(strict_types=1);
```

File: Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:45]: The use of the nodetypes package is depreacated. We should not require it.

File: Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:67]: Please add proper typing throughout the classes

File: Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php:4]: Please keep the license comment and add it also to the newly created files

File: Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:186]: Looks like a method that should be put to an own service / utility class

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php
Comment (daniellienert) [inline Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:227]: IMHO the better way is to define a `SiteGeneratorInterface` and select classes implementing this interface

File: Neos.SiteKickstarter/Classes/Service/SitePackageGeneratorNameService.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Service/SitePackageGeneratorNameService.php:20]: Honestly I don't like the usage of special annotations here. Wouldn't it be more clear to add `getGeneratorName` / `getGeneratorDescription` to the generatorInterface or use just the className.

File: Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php
Comment (daniellienert) [inline Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php:18]: Should we really offer a Fluid template generator? I guess when asked, we recommend not to use fluid for site templating anymore.

**Review by daniellienert — CHANGES_REQUESTED** [review]
Hey @Wyndala,
thanks a lot for your work! I added some comments - lots of nitpicks and some serious complains:
- I would not recommend to use annotations just for a display name, there are more straight-forward approaches. (but I am open to discuss that)
- I would like to discuss if we really should offer a fluid generator, we do recommend to not use fluid anymore

Cheers, 
Daniel

**Comment by bwaidelich** [comment]
Great suggestions by Daniel and I agree to his comments

**Comment by Wyndala** [comment]
Thx for the feedback -> I try to put it into the pr in this week :).

**Comment by bwaidelich** [comment]
Update: I had created an issue (#3099) for that since the PR wasn't linked to a ticket. But in fact there _was_ an issue already (#3074) so I updated the description accordingly.

@Wyndala any news on this one?

**Comment by jonnitto** [comment]
@Wyndala & @ahaeslich The feature freeze will be on the 9th of November, so it would be really awesome if you could finish this PR. If you have questions, don't wait to ask 😄

**Comment by Wyndala** [comment]
@jonnitto I am currently on it :). I am currently testing my stuff. I try my best to get it work on the weekend.

**Comment by Wyndala** [comment]
Hey @jonnitto, @daniellienert, @bwaidelich,

I just finished adding all the feedback.

I extracted the FluidTemplateGenerator to https://github.com/queoGmbH/FluidSiteGenerator . So in case somebody needs something like this, the code is not lost (even I don't like fluid and it is deprecated).

I tested the stuff in my biggest NEOS Installation and it worked fine. If there is something else I can do, just leave feedback again. After month of hard work my week looks a little bit better the next months. :)

Have a nice sunday,
Ralf

File: Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion
Comment (jonnitto) [inline Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:52]:
```suggestion
    javascripts = Neos.Fusion:Join
```
Make usage of `Neos.Fusion:Join` and indent with spaces

File: Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion
Comment (jonnitto) [inline Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:54]:
```suggestion
    @process.appendJavaScripts = ${value + this.javascripts}
```
Indent with spaces

File: Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion
Comment (jonnitto) [inline Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:57]:
```suggestion
  bodyJavaScripts = Neos.Fusion:Join
```

File: Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion
Comment (jonnitto) [inline Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:23]:
```suggestion
     stylesheets.site = afx`
        <link rel="stylesheet" href={StaticResource.uri('{packageKey}', 'Public/site.css')} />
    `
```
Since we have now an [eel helper for static sources](https://github.com/neos/neos-development-collection/pull/3134), we could make use of this.

File: Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion
Comment (jonnitto) [inline Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:33]:
```suggestion
    javascripts.site = afx`
        <link type="text/javascript" defer href={StaticResource.uri('{packageKey}', 'Public/site.js')} />
    `
```
We could also use here the new Eel helper. And I suggest also to add the defer attribute…

**Review by jonnitto — CHANGES_REQUESTED** [review]
I tested your PR and it works fine! ❤️ 
I made a few suggestions about the generated fusion code

**Review by bwaidelich — COMMENTED** [review]
The change set still contains two empty Doctrine Migrations (Neos.SiteKickstarter/Migrations/Code/Version20161125002311.php and Neos.SiteKickstarter/Migrations/Code/Version20161125095901.php) that break the CI

**Comment by jonnitto** [comment]
> The change set still contains two empty Doctrine Migrations (Neos.SiteKickstarter/Migrations/Code/Version20161125002311.php and Neos.SiteKickstarter/Migrations/Code/Version20161125095901.php) that break the CI

@bwaidelich Are you sure? I checked out this PR and these files have content. As I see, the filemode changed...

**Comment by Wyndala** [comment]
Hey guys,

I fixed the weird file issue and now the ci is green. :)

I added jon's feedback and I think we are ready to merge (from my point of view) :)

**Review by jonnitto — APPROVED** [review]

**Review by daniellienert — COMMENTED** [review]
Hey @Wyndala, looks much better now. Thanks a lot. 

I wasn't able to get it to work though, I get the error 

```
./flow kickstart:site Foo.Bar barfoo
Notice: Trying to access array offset on value of type null in /Users/ry25/Workspace/Projects/neosdev/Packages/Framework/Neos.Flow/Classes/Package/PackageManager.php line 373
```
Not sure if its related to your changes. @jonnitto have you tested it, does it work for you?

**Comment by jonnitto** [comment]
> flow kickstart:site Foo.Bar barfoo

This works for me, yes. Tested it also without arguments.

**Comment by Wyndala** [comment]
> > flow kickstart:site Foo.Bar barfoo
> 
> This works for me, yes. Tested it also without arguments.

I also tested it again on a fresh installation and it worked without any error

**Comment by bwaidelich** [comment]
@daniellienert any reason still to block this one from your side?

**Comment by daniellienert** [comment]
@bwaidelich, there are still parts of the code which would say is a bit over-engineered / bulky which I at least would discuss if there was time to do so, like the `SiteGeneratorCollectingService` or the `SitePackageGeneratorNameService`. But if you think its fine, Its fine for me.

**Comment by bwaidelich** [comment]
Thanks for the feedback

> But if you think its fine, Its fine for me.

I didn't really get into it very deeply. In general I think it's a great improvement feature-wise and it's not userland code anyways but "just" a tool for devs. So I wouldn't be too concerned about architecture in that part. Also we still have ~3 weeks to fix potential bugs.

Having said that: Since I didn't really test/evaluate this, I'll abstain from voting  and leave it up to @jonnitto to decide

**Comment by jonnitto** [comment]
I can't judge over the PHP part as @daniellienert, but as we say it is best practice to use `afx`, I think it is a really great feature to get rid of the old style.

**Pull request merged**
